azero/contracts/money-box/contract/lib.rs

```diff
@@ -0,0 +1,233 @@
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+pub use self::money_box_contract::MoneyBoxContractRef;
+
+#[ink::contract]
+pub mod money_box_contract {
+    use ink::{
+        env::{set_code_hash, Error as InkEnvError},
+        prelude::{format, string::String},
+    };
+    use money_box_trait::MoneyBox;
+    use scale::{Decode, Encode};
+
+    #[ink(event)]
+    #[derive(Debug)]
+    #[cfg_attr(feature = "std", derive(Eq, PartialEq))]
+    pub struct PocketMoneyPaidOut {
+        #[ink(topic)]
+        pub to: AccountId,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    #[cfg_attr(feature = "std", derive(Eq, PartialEq))]
+    pub struct InsufficientFunds {
+        pub current_balance: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    #[cfg_attr(feature = "std", derive(Eq, PartialEq))]
+    pub struct ContractUpgraded {
+        pub new_code_hash: [u8; 32],
+    }
+
+    #[derive(Debug, PartialEq, Eq, Encode, Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum MoneyBoxError {
+        CallerNotAdmin,
+        InkEnvError(String),
+    }
+
+    impl From<InkEnvError> for MoneyBoxError {
+        fn from(why: InkEnvError) -> Self {
+            Self::InkEnvError(format!("{:?}", why))
+        }
+    }
+
+    #[ink(storage)]
+    pub struct MoneyBoxContract {
+        amount_to_pay: Balance,
+        owner: AccountId,
+        admin: AccountId,
+    }
+
+    impl MoneyBoxContract {
+        #[ink(constructor)]
+        pub fn new(amount_to_pay: Balance) -> Self {
+            Self {
+                amount_to_pay,
+                owner: Self::env().caller(),
+                admin: Self::env().caller(),
```

Comment:
instead of two fields and duplicated  ensure_a, ensure_b etc introduce an e.g. `Role` enum and a mapping AccountId -> Role. You can then have one common check function one query function and one update function